### PR TITLE
Feature/apparatus entries box

### PR DESCRIPTION
**Pull request by giuliac89**

### CODE REVIEW

File: src/app/components/apparatus-entry/apparatus-entry-detail/apparatus-entry-detail.component.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  get significantRdg(): Reading[] {
+    return this.data.readings.filter((rdg) => rdg.significant);
+  }
+
+  get notSignificantRdg(): Reading[] {
+    return this.data.readings.filter((rdg) => !rdg.significant);
+  }
+
+  get readings(): Reading[] {
+    return [this.data.lemma].concat(this.significantRdg).concat(this.notSignificantRdg);
```
Comment (szenzaro) [inline src/app/components/apparatus-entry/apparatus-entry-detail/apparatus-entry-detail.component.ts:41]: `return [this.data.lemma, ...this.significantRdg, ...this.notSignificantRdg]`

File: src/app/components/apparatus-entry/apparatus-entry-detail/apparatus-entry-detail.component.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+import { EVTModelService } from '../../../services/evt-model.service';
+@Component({
+  selector: 'evt-apparatus-entry-detail',
+  templateUrl: './apparatus-entry-detail.component.html',
+  styleUrls: ['./apparatus-entry-detail.component.scss'],
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+
+@register(ApparatusEntryDetailComponent)
+export class ApparatusEntryDetailComponent {
```
Comment (szenzaro) [inline src/app/components/apparatus-entry/apparatus-entry-detail/apparatus-entry-detail.component.ts:13]: this seems to contain duplicated code from the app/rdg parsers. If so, try to refactor this.

File: src/app/components/apparatus-entry/apparatus-entry-detail/apparatus-entry-detail.component.html [hunk 3]
```diff
@@ -0,0 +1,104 @@
+<div class="card app-detail-container">
```
Comment (szenzaro) [inline src/app/components/apparatus-entry/apparatus-entry-detail/apparatus-entry-detail.component.html:1]: This component is very large. Is it possible to modularize it a bit?

File: src/app/components/apparatus-entry/apparatus-entry.component.ts [hunk 4]
```diff
@@ -1,26 +1,59 @@
...
+  public nestedApps: ApparatusEntry[] = [];
+
   variance$ = this.evtModelService.appVariance$.pipe(
     map((variances) => variances[this.data.id]),
+    shareReplay(1),
   );
 
   constructor(
     private evtModelService: EVTModelService,
+    @Optional() @SkipSelf() private parentDetailComponent?: ApparatusEntryDetailComponent,
```
Comment (szenzaro) [inline src/app/components/apparatus-entry/apparatus-entry.component.ts:28]: Are these decorators needed?

File: src/app/components/apparatus-entry/apparatus-entry.component.ts [hunk 5]
```diff
@@ -1,26 +1,59 @@
...
   ) {
-    super();
+    this.isInsideAppDetail = this.parentDetailComponent ? true : false;
+  }
+
+  ngOnInit() {
+    if (this.data.hasNestedApp) {
+      this.getNestedApps(this.data);
```
Comment (szenzaro) [inline src/app/components/apparatus-entry/apparatus-entry.component.ts:35]: The name `getNestedApps` suggests that some value is returned. Instead it is a procedure.

File: src/app/components/apparatus-entry/apparatus-entry.component.ts [hunk 6]
```diff
@@ -1,26 +1,59 @@
...
+    });
+  }
+
+  toggleAppEntryBox(e: MouseEvent) {
+    e.stopPropagation();
+    this.opened = !this.opened;
+  }
+
+  closeAppEntryBox() {
+    if (this.opened) {
```
Comment (szenzaro) [inline src/app/components/apparatus-entry/apparatus-entry.component.ts:55]: you can just set it to false since if this.opened is true it will change to false, otherwise it will remain false

File: src/app/models/evt-models.ts [hunk 7]
```diff
@@ -159,12 +158,15 @@ export interface WitnessGroup {
...
     groupId: string;
 }
+
 export class ApparatusEntry extends GenericElement {
     id: string;
     lemma: Reading;
     readings: Reading[];
     notes: Note[];
     originalEncoding: string;
+    hasNestedApp: boolean;
```
Comment (szenzaro) [inline src/app/models/evt-models.ts:168]: Isn't this equivalent to `nestedAppsIDs.length`?

**Review by szenzaro — CHANGES_REQUESTED** [review]
Missing ChangeLog

**Review by ChiaraDipi — CHANGES_REQUESTED** [review]
I agree with szenzaro comments and I would also try to modularize the ApparatusEntryDetail component if possible

File: src/app/components/apparatus-entry/apparatus-entry.component.ts (hunk 4 above)
Comment (giuliac89) [reply src/app/components/apparatus-entry/apparatus-entry.component.ts:28]: @Optional() it's necessary because normally an `apparatusEntry` doesn't have a parent which is an `ApparatusEntryDetailComponent`. This only applies to nested apps.
@SkipSelf seems useless (I can't understand why I added it)

File: src/app/components/apparatus-entry/apparatus-entry-detail/apparatus-entry-detail.component.html (hunk 3 above)
Comment (ChiaraDipi) [reply src/app/components/apparatus-entry/apparatus-entry-detail/apparatus-entry-detail.component.html:1]: I would manage modularization in a new distinct feature.

File: src/app/components/apparatus-entry/apparatus-entry-detail/apparatus-entry-detail.component.ts [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+  }
+
+  get rdgMetadata() {
+    return Object.keys(this.data.attributes).filter((key) => key !== 'id')
+      .reduce((obj, key) => {
+        obj[key] = this.data.attributes[key];
+
+        return obj;
+      },      {});
+  }
```
Comment (ChiaraDipi) [inline src/app/components/apparatus-entry/apparatus-entry-detail/apparatus-entry-detail.component.ts:37]: I think it can be changed into 
```
return Object.keys(this.data.attributes).filter((key) => key !== 'id')
  .reduce(
    (obj, key) => ({
      ...obj,
      [key]: this.data.attributes[key],
    }),
    {});
```

File: src/app/components/apparatus-entry/apparatus-entry-detail/apparatus-entry-detail.component.ts [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+
+  getNestedAppLemma(appId: string): Reading {
+    return this.nestedApps.find((c) => c.id === appId).lemma;
+  }
+
+  getNestedAppPos(appId: string): number {
+    const currentApp = this.nestedApps.find(nesApp => nesApp.id === appId);
+
+    return this.nestedApps.indexOf(currentApp);
```
Comment (ChiaraDipi) [inline src/app/components/apparatus-entry/apparatus-entry-detail/apparatus-entry-detail.component.ts:71]: Try using directly 
```
this.nestedApps.findIndex(nesApp => nesApp.id === appId);
```

File: src/app/components/apparatus-entry/apparatus-entry-readings/apparatus-entry-readings.component.ts [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+  get significantRdg(): Reading[] {
+    return this.data.readings.filter((rdg) => rdg.significant);
+  }
+
+  getWits$(witID: string): Observable<string[]> {
+    return this.groups$.pipe(
+      map((groups) => {
+        const currentGroup = groups.filter((g) => g.id === witID);
+
+        return currentGroup.map((g) => g.witnesses).reduce((x, y) => ([ ...x, ...y ]), []);
```
Comment (ChiaraDipi) [inline src/app/components/apparatus-entry/apparatus-entry-readings/apparatus-entry-readings.component.ts:38]: Since it is an array you can name the variable "plural": `currentGroups`. Otherwise you can chain everything in one single line:
```
map((groups) =>
  groups.filter((g) => g.id === witID).map((g) => g.witnesses).reduce((x, y) => ([...x, ...y]), [])
),
```

File: src/app/services/xml-parsers/app-parser.ts [hunk 11]
```diff
@@ -52,8 +52,7 @@ export class RdgParser extends EmptyParser implements Parser<XMLElement> {
         return !Array.from(attributes).some(({ name, value }) => notSignificantReading.includes(`${name}=${value}`));
     }
 }
-
-@xmlParser('app', AppParser)
+@xmlParser('evt-apparatus-entry-parser', AppParser)
```
Comment (ChiaraDipi) [inline src/app/services/xml-parsers/app-parser.ts:55]: Why this change in tagName definition?

**Review by ChiaraDipi — CHANGES_REQUESTED** [review]

File: src/app/services/xml-parsers/app-parser.ts (hunk 11 above)
Comment (giuliac89) [reply src/app/services/xml-parsers/app-parser.ts:55]: @ChiaraDipi I think I changed it to use the new `ParserRegister.get` function. See commit 6a6494d

**Review by ChiaraDipi — APPROVED** [review]
